// @spheron/core/dist/spheron-api/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeTypeEnum = exports.ClusterProtocolEnum = exports.PersistentStorageClassEnum = exports.InstanceLogType = exports.MarketplaceCategoryEnum = exports.HealthStatusEnum = exports.InstanceStateEnum = exports.ClusterStateEnum = exports.ProviderEnum = exports.AppTypeEnum = exports.DeploymentStatusEnum = exports.DomainApplicationTypeEnum = exports.DomainTypeEnum = exports.FrameworkEnum = exports.NodeVersionEnum = exports.ProjectStateEnum = exports.DeploymentEnvironmentStatusEnum = exports.ProjectTypeEnum = exports.ProtocolEnum = void 0;
var ProtocolEnum;
(function (ProtocolEnum) {
    ProtocolEnum["ARWEAVE"] = "arweave";
    ProtocolEnum["FILECOIN"] = "ipfs-filecoin";
    ProtocolEnum["IPFS"] = "ipfs";
})(ProtocolEnum = exports.ProtocolEnum || (exports.ProtocolEnum = {}));
var ProjectTypeEnum;
(function (ProjectTypeEnum) {
    ProjectTypeEnum["STANDARD"] = "STANDARD";
    ProjectTypeEnum["UPLOAD"] = "UPLOAD";
})(ProjectTypeEnum || (ProjectTypeEnum = {}));
exports.ProjectTypeEnum = ProjectTypeEnum;
var DeploymentEnvironmentStatusEnum;
(function (DeploymentEnvironmentStatusEnum) {
    DeploymentEnvironmentStatusEnum["ACTIVE"] = "ACTIVE";
    DeploymentEnvironmentStatusEnum["INACTIVE"] = "INACTIVE";
    DeploymentEnvironmentStatusEnum["PAYMENT_PENDING"] = "PAYMENT_PENDING";
})(DeploymentEnvironmentStatusEnum || (DeploymentEnvironmentStatusEnum = {}));
exports.DeploymentEnvironmentStatusEnum = DeploymentEnvironmentStatusEnum;
var ProjectStateEnum;
(function (ProjectStateEnum) {
    ProjectStateEnum["MAINTAINED"] = "MAINTAINED";
    ProjectStateEnum["ARCHIVED"] = "ARCHIVED";
})(ProjectStateEnum || (ProjectStateEnum = {}));
exports.ProjectStateEnum = ProjectStateEnum;
var DomainTypeEnum;
(function (DomainTypeEnum) {
    DomainTypeEnum["DOMAIN"] = "domain";
    DomainTypeEnum["SUBDOMAIN"] = "subdomain";
    DomainTypeEnum["HANDSHAKE_DOMAIN"] = "handshake-domain";
    DomainTypeEnum["HANDSHAKE_SUBDOMAIN"] = "handshake-subdomain";
    DomainTypeEnum["ENS_DOMAIN"] = "ens-domain";
})(DomainTypeEnum || (DomainTypeEnum = {}));
exports.DomainTypeEnum = DomainTypeEnum;
var DomainApplicationTypeEnum;
(function (DomainApplicationTypeEnum) {
    DomainApplicationTypeEnum[DomainApplicationTypeEnum["WEB"] = 1] = "WEB";
    DomainApplicationTypeEnum[DomainApplicationTypeEnum["COMPUTE"] = 2] = "COMPUTE";
})(DomainApplicationTypeEnum || (DomainApplicationTypeEnum = {}));
exports.DomainApplicationTypeEnum = DomainApplicationTypeEnum;
var NodeVersionEnum;
(function (NodeVersionEnum) {
    NodeVersionEnum["V_12"] = "V_12";
    NodeVersionEnum["V_14"] = "V_14";
    NodeVersionEnum["V_16"] = "V_16";
})(NodeVersionEnum || (NodeVersionEnum = {}));
exports.NodeVersionEnum = NodeVersionEnum;
var FrameworkEnum;
(function (FrameworkEnum) {
    FrameworkEnum["SIMPLE_JAVASCRIPT_APP"] = "static";
    FrameworkEnum["VUE"] = "vue";
    FrameworkEnum["REACT"] = "react";
    FrameworkEnum["NEXT"] = "next";
    FrameworkEnum["ANGULAR"] = "angular";
    FrameworkEnum["PREACT"] = "preact";
    FrameworkEnum["NUXT"] = "nuxt2";
    FrameworkEnum["SVELTE"] = "svelte";
    FrameworkEnum["GATSBY"] = "gatsby";
    FrameworkEnum["ELEVENTY"] = "eleventy";
    FrameworkEnum["DOCUSAURUS"] = "docusaurus";
    FrameworkEnum["SANITY"] = "sanity";
    FrameworkEnum["HUGO"] = "hugo";
    FrameworkEnum["IONIC_REACT"] = "ionicreact";
    FrameworkEnum["VITE"] = "vite";
    FrameworkEnum["SCULLY"] = "scully";
    FrameworkEnum["STENCIL"] = "stencil";
    FrameworkEnum["BRUNCH"] = "brunch";
    FrameworkEnum["IONIC_ANGULAR"] = "ionicangular";
})(FrameworkEnum || (FrameworkEnum = {}));
exports.FrameworkEnum = FrameworkEnum;
var DeploymentStatusEnum;
(function (DeploymentStatusEnum) {
    DeploymentStatusEnum["PRE_QUEUE"] = "PreQueue";
    DeploymentStatusEnum["QUEUED"] = "Queued";
    DeploymentStatusEnum["PENDING"] = "Pending";
    DeploymentStatusEnum["CANCELED"] = "Canceled";
    DeploymentStatusEnum["DEPLOYED"] = "Deployed";
    DeploymentStatusEnum["FAILED"] = "Failed";
    DeploymentStatusEnum["AUTHORIZATION_NEEDED"] = "AuthorizationNeeded";
    DeploymentStatusEnum["KILLING"] = "Killing";
    DeploymentStatusEnum["TIMED_OUT"] = "TimedOut";
})(DeploymentStatusEnum || (DeploymentStatusEnum = {}));
exports.DeploymentStatusEnum = DeploymentStatusEnum;
var AppTypeEnum;
(function (AppTypeEnum) {
    AppTypeEnum["WEB_APP"] = "app";
    AppTypeEnum["COMPUTE"] = "compute";
    AppTypeEnum["STORAGE"] = "storage";
})(AppTypeEnum || (AppTypeEnum = {}));
exports.AppTypeEnum = AppTypeEnum;
var ProviderEnum;
(function (ProviderEnum) {
    ProviderEnum["GITHUB"] = "GITHUB";
    ProviderEnum["GITLAB"] = "GITLAB";
    ProviderEnum["BITBUCKET"] = "BITBUCKET";
    ProviderEnum["DOCKERHUB"] = "DOCKERHUB";
})(ProviderEnum || (ProviderEnum = {}));
exports.ProviderEnum = ProviderEnum;
var ClusterStateEnum;
(function (ClusterStateEnum) {
    ClusterStateEnum["MAINTAINED"] = "MAINTAINED";
    ClusterStateEnum["ARCHIVED"] = "ARCHIVED";
})(ClusterStateEnum || (ClusterStateEnum = {}));
exports.ClusterStateEnum = ClusterStateEnum;
var InstanceStateEnum;
(function (InstanceStateEnum) {
    InstanceStateEnum["STARTING"] = "Starting";
    InstanceStateEnum["FAILED_START"] = "Failed-start";
    InstanceStateEnum["ACTIVE"] = "Active";
    InstanceStateEnum["CLOSED"] = "Closed";
})(InstanceStateEnum || (InstanceStateEnum = {}));
exports.InstanceStateEnum = InstanceStateEnum;
var HealthStatusEnum;
(function (HealthStatusEnum) {
    HealthStatusEnum["ACTIVE"] = "Active";
    HealthStatusEnum["INACTIVE"] = "Inactive";
    HealthStatusEnum["UNKNOWN"] = "Unknown";
})(HealthStatusEnum || (HealthStatusEnum = {}));
exports.HealthStatusEnum = HealthStatusEnum;
var MarketplaceCategoryEnum;
(function (MarketplaceCategoryEnum) {
    MarketplaceCategoryEnum["DATABASE"] = "Database";
    MarketplaceCategoryEnum["NODE"] = "Node";
    MarketplaceCategoryEnum["TOOLS"] = "Tools";
})(MarketplaceCategoryEnum || (MarketplaceCategoryEnum = {}));
exports.MarketplaceCategoryEnum = MarketplaceCategoryEnum;
var InstanceLogType;
(function (InstanceLogType) {
    InstanceLogType["DEPLOYMENT_LOGS"] = "logs";
    InstanceLogType["INSTANCE_LOGS"] = "instanceLogs";
    InstanceLogType["INSTANCE_EVENTS"] = "instanceEvents";
})(InstanceLogType || (InstanceLogType = {}));
exports.InstanceLogType = InstanceLogType;
var PersistentStorageClassEnum;
(function (PersistentStorageClassEnum) {
    PersistentStorageClassEnum["HDD"] = "beta1";
    PersistentStorageClassEnum["SSD"] = "beta2";
    PersistentStorageClassEnum["NVMe"] = "beta3";
})(PersistentStorageClassEnum || (PersistentStorageClassEnum = {}));
exports.PersistentStorageClassEnum = PersistentStorageClassEnum;
var ClusterProtocolEnum;
(function (ClusterProtocolEnum) {
    ClusterProtocolEnum["AKASH"] = "akash";
})(ClusterProtocolEnum || (ClusterProtocolEnum = {}));
exports.ClusterProtocolEnum = ClusterProtocolEnum;
var ComputeTypeEnum;
(function (ComputeTypeEnum) {
    ComputeTypeEnum["SPOT"] = "spot";
    ComputeTypeEnum["DEMAND"] = "demand";
})(ComputeTypeEnum || (ComputeTypeEnum = {}));
exports.ComputeTypeEnum = ComputeTypeEnum;
